src/core/src/engine/consensus_state.rs

```diff
@@ -137,110 +104,203 @@ struct Entry {
+                warn!("Wait certificate is empty"); // non-poet block
+                if blocks.is_empty() || prev_wait_cert.is_some() {
+                    blocks.push(Entry {
+                        key: cur_block_id.clone(),
+                        value: BlockInfo {
+                            wait_certificate: None,
+                            validator_info: None,
+                        },
+                    });
+                }
+            } else {
+                 let cur_block_validator_id = poet2_util::to_hex_string(&cur_block.signer_id);
+                 // Get validator info for validator_id at current block_id
+                 let cur_validator_info = validator_registry_view::get_validator_info_for_validator_id(cur_block_validator_id.as_str(), &cur_block_id.to_owned(), poet2_service);
+                 if cur_validator_info.is_err() {
+                     debug!("Cannot find registry entry for block_id: {}", poet2_util::to_hex_string(&cur_block_id));
+                 }
+                 debug!("Building consensus state for block_id: {}", poet2_util::to_hex_string(&cur_block_id));
+
+                 blocks.push(Entry {
+                     key: cur_block_id.clone(),
+                     value: BlockInfo {
+                         wait_certificate: Some(cur_wait_cert.clone()),
+                         validator_info: Some(cur_validator_info.unwrap().clone()),
+                     },
+                 });
             }
-            previous_wait_certificate = wait_certificate.clone();
-            current_id = block.previous_id;
-            //let mut consensus_state = consensus_state_store_.get(current_id.clone());
-            if consensus_state.is_none() {
-                consensus_state = Some(ConsensusState::default());
+
+            if cur_block.block_num == 0 {
+                debug!("Reached genesis! Breaking out of loop...");
+                break;
             }
-            for entry in blocks.iter().rev() {
-                let mut val = &entry.value;
-                if val.wait_certificate.is_none() {
-                    consensus_state = Some(ConsensusState::default());
-                } else {
-                    self.validator_did_claim_block(
-                        &(val.clone().validator_info.unwrap()),
-                        &(val.clone().wait_certificate.unwrap()),
-                        &(val.clone().poet_settings_view.unwrap()),
-                    );
+
+            prev_wait_cert = Some(cur_wait_cert.clone());
+
+
+            // Set cur_block id to cur_block's previous block id for next iteration
+            let prev_block_id = cur_block.previous_id.clone();
+
+            // Set cur_block
+            // Find the previous block in block cache
+
+            warn!("Finding block for current blocks previous block id");
+            // find from service
+            cur_block = poet2_service.get_block(&prev_block_id).expect("Failed to get a block!");
```

Comment:
Why not pass the blocks obtained during fork resolution? Use service only for fall back.